File: validation/cases/care_solution/care_solution.cpp
// care_solution.cpp -- CARE solution matching care_solution.m
// Usage: ./care_solution > care_solution_cpp.csv

#include "ctrlpp/control/care.h"
#include "ctrlpp/control/lqr.h"

#include <cstdio>

int main()
{
    using Scalar = double;
    constexpr std::size_t NX = 2;
    constexpr std::size_t NU = 1;

    Eigen::Matrix<Scalar, NX, NX> A;
    A << 0.0, 1.0, -1.0, -0.5;
    Eigen::Matrix<Scalar, NX, NU> B;
    B << 0.0, 1.0;

    Eigen::Matrix<Scalar, NX, NX> Q;
    Q << 10.0, 0.0, 0.0, 1.0;
    Eigen::Matrix<Scalar, NU, NU> R;
    R << 1.0;

    auto P = ctrlpp::care<Scalar, NX, NU>(A, B, Q, R).value().P;
    auto K = ctrlpp::lqr_gain_continuous<Scalar, NX, NU>(A, B, Q, R).value();

    std::printf("P_00,P_01,P_10,P_11,K_00,K_01\n");
    std::printf("%.15e,%.15e,%.15e,%.15e,%.15e,%.15e\n",
                P(0, 0), P(0, 1), P(1, 0), P(1, 1), K(0, 0), K(0, 1));
}

File: validation/cases/care_solution/care_solution.m
% care_solution.m -- Reference CARE solution via Octave Control Toolbox.
% Outputs P (row-major) and K (row-major) as a single CSV row.
%
% Usage: octave --no-gui care_solution.m > care_solution_octave.csv

pkg load control;

% Mass-spring-damper: m=1, k=1, b=0.5
% Continuous: A_c = [0 1; -1 -0.5], B_c = [0; 1].
% Solve continuous-time CARE A'P + PA - PBR^{-1}B'P + Q = 0 directly.

A = [0, 1; -1, -0.5];
B = [0; 1];
Q = [10 0; 0 1];
R = [1];

[P, ~, ~] = care(A, B, Q, R);

% Continuous-time LQR gain K = R^{-1} B' P
K = R \ (B' * P);

fprintf('P_00,P_01,P_10,P_11,K_00,K_01\n');
fprintf('%.15e,%.15e,%.15e,%.15e,%.15e,%.15e\n', ...
        P(1,1), P(1,2), P(2,1), P(2,2), K(1,1), K(1,2));
